### ISSUE TITLE
Fix authentication for https endpoint on kubelets

### ISSUE THREAD

**Issue by StianOvrevage** [Azure/AKS#680]
**What happened**:

Installed `prometheus-operator` and `kube-prometheus` helm charts.

This creates among other things a ServiceMonitor for gathering metrics from the kubelets running on the nodes installed by AKS.

One of the entries look like this:
```
  - bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
    interval: 15s
    port: https-metrics
    scheme: https
    tlsConfig:
      caFile: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
      insecureSkipVerify: true
```

However that gives a 401 Unauthorized in Prometheus.

A known workaround is to patch the ServiceMonitor so that it uses http instead:
```
  - bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
    interval: 15s
    port: http-metrics
    scheme: http
```

It would be nice to not have to manually add this patch every time we install prometheus on a new cluster (which we do daily).

**So in short: Fix the https metrics endpoint on kubelets to accept the same auth as the http endpoint.**

**What you expected to happen**:

Be able to view kubelet metrics without any changes.

**How to reproduce it (as minimally and precisely as possible)**:

Use helm to install prometheus-operator and kube-prometheus. View targets in Prometheus UI.

**Environment**:
- Kubernetes version (use `kubectl version`): 1.11.2
- Size of cluster (how many worker nodes are in the cluster?) 3

**Comment by tomasr** [comment]
You can add that service monitor configuration directly in the prometheus chart configuration before feeding it to Helm. It's what I do:

```yaml
prometheus:
  ingress:
    enabled: false
  retention: 168h
  ruleNamespaceSelector: {}
  rulesSelector: {}
  rules:
    specifiedInValues: true
    additionalLabels: {}
    value: {}
...
  serviceMonitors:
    # override the default kubelet monitor
    # to use HTTP and not HTTPS
    - name: kubelet
      selector:
        matchLabels:
          k8s-app: kubelet
      namespaceSelector:
        matchNames:
          - kube-system
      endpoints:
        - port: http-metrics
          scheme: http
          interval: 30s
          bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
        - port: http-metrics
          scheme: http
          path: /metrics/cadvisor
          interval: 30s
          bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
  storageSpec:
...
```

**Comment by andig** [comment]
I'm using the stable `prometheus-operator` helm chart. Ootb, `http-metrics` already works, `cadvisor` doesn't. Chaning the `servicemontor`like @tomasr doesn't work for me. With

	spec:
	  endpoints:
	  - honorLabels: true
	    interval: 30s
	    port: http-metrics
	  - honorLabels: true
	    interval: 30s
	    path: /metrics/cadvisor
	    port: http-metrics

`cadvisor` is still offline?

**Comment by andig** [comment]
Unfortunately I'm not able to deploy this automatically:

	kubectl patch servicemonitor -n kube-system prometheus-prometheus-oper-kubelet --patch "$(cat prometheus/kubelet-cadvisor-patch.yaml)"

	Error from server (UnsupportedMediaType): the body of the request was in an unknown format - accepted media types include: application/json-patch+json, application/merge-patch+json

**Comment by larsduelfer** [comment]
Patches are ok, but the root cause should be fixed!

**Comment by cmendible** [comment]
The solution by @tomasr worked for me...

**Comment by jnoller** [comment]
Closing this issue as old/stale.

If this issue still comes up, please confirm you are running the latest AKS release. If you are on the latest release and the issue can be re-created outside of your specific cluster please open a new github issue.

If you are only seeing this behavior on clusters with a unique configuration (such as custom DNS/VNet/etc) please open an Azure technical support ticket.
